cilium-cli: Validate NodePort readiness on all node IP addresses

pillai-ashwin opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer's Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

## Description

Fixes intermittent `pod-to-local-nodeport` connectivity test failures (exit code 28 - connection timeout) in dual-stack clusters.

### Root Cause

The upfront NodePort readiness check in `deployment.go` only validated `HostIP` (one IP per node, typically IPv4), but tests iterate through all addresses in `node.Status.Addresses[]` which includes both IPv4 and IPv6 in dual-stack clusters.

This meant IPv6 NodePorts were never validated for readiness, causing timeouts when tests tried them.

### Fix

Enhance the upfront NodePort validation to check all node IP addresses (NodeInternalIP and NodeExternalIP types) instead of just HostIP. This ensures all addresses that will be tested are validated during setup.

Fixes: #39793

```release-note
Fix intermittent NodePort connectivity test timeouts in dual-stack clusters by validating NodePort readiness on all node IP addresses during test setup.
```

dylandreimerink commented on cilium-cli/connectivity/tests/service.go:
```diff
@@ -248,7 +248,34 @@ func curlNodePort(ctx context.Context, s check.Scenario, t *check.Test,
+				// Add detailed logging for IPv6 NodePort connections
+				if ipFam == features.IPFamilyV6 {
+					a.Debugf("Testing IPv6 NodePort connectivity: %s -> %s:%d", pod.Name(), addr.Address, np)
+				}
```
Adding logs like this is fine during an investigation, in like a draft PR. But I do not thing we should commit them into the main branch to be part of the tools everyone will use.

pillai-ashwin replied:
@dylandreimerink Thanks a lot for your detailed feedback! 😊 I've made some changes based on that -
- Yes to your comment on timeouts and masking risk. Was just trying to reason with different hypothetical reasons (some of them far from reality haha) I removed IPv6-specific timeout multipliers and kept the default 2s connect-timeout uniformly across IP families. The fix now focuses on gating curl by explicit NodePort readiness rather than extending timeouts.

- In agreement on retries too. I removed default retries from NodePort actions so failures after readiness will hard-fail instead of being retried.

- Implemented readiness gating. The tests now wait for NodePort availability (`WaitForNodePorts`) before curl, which addresses the flake without relaxing timing. The `no-policies-extra/pod-to-local-nodeport` scenario runs consistently under this change.

dylandreimerink reviewed (changes requested):
> 1. Insufficient timeouts: 2-second connect timeout was too aggressive for IPv6 NodePort connections
> 2. IPv6 networking delays: IPv6 connections can take longer to establish than IPv4

I am a bit skeptical about these findings. Establishing a connection should not take 2 seconds, especially when all of it is effective on the same node (no internet involved). If they do, then it seems to me that something is going wrong (packets being dropped, causing TCP level retires, or something related) and increasing the timeout may just mask that.

I also have never heard of IPv6 being slower than IPv4 to establish connections. I don't think we should accept that statement without something to back that up.

> Missing retry logic: Single-attempt connections failed on transient network issues

This is essentially saying that we expect Cilium, the thing we are testing to fail. I don't think that is right. These sorts of tests are exactly here to catch issues in Cilium. If the first connection we make always fails or frequently fails then we should investigate why and not plaster over it with retries. We want all connections to work, not just at least one out of three.

> Premature testing: Tests started before NodePorts were fully ready

This sounds more like the right direction. I think this is more likely to be the real fix.

viktor-kurchenko reviewed (approved):
Thank you!

joestringer commented on cilium-cli/connectivity/tests/service.go:
```diff
@@ -251,6 +251,11 @@ func curlNodePort(ctx context.Context, s check.Scenario, t *check.Test,
 			// Create the Action with the original svc as this will influence what the
 			// flow matcher looks for in the flow logs.
 			t.NewAction(s, name, pod, svc, features.IPFamilyAny).Run(func(a *check.Action) {
+				// Wait for NodePort to be ready before testing connectivity
```
Why aren't all the nodeports checked to be ready right at the start of the full connectivity suite run? Do we need to perform this step each time we run the tests?

pillai-ashwin committed:
cilium-cli: Validate NodePort readiness on all node IP addresses

The upfront NodePort readiness check only validated HostIP (typically one
IP per node), but tests target all addresses in node.Status.Addresses which
can include multiple IPs per node (e.g., IPv4 and IPv6 in dual-stack clusters).

This caused intermittent failures (exit code 28) when tests tried IPv6 NodePorts
that were never validated for readiness.

Fix by enhancing the upfront validation in deployment.go to check NodePort
readiness on all node IP addresses (InternalIP and ExternalIP types) during
the setup phase, ensuring all addresses that will be tested are ready before
tests execute.

Fixes: #39793
Signed-off-by: Ashwin Pillai <pillaiashwin96@gmail.com>
